### DISCUSSION TITLE
fit two data sets using multitasks Multi-Variate Gaussian Process

### DISCUSSION THREAD

**Discussion by lovecodng** [cornellius-gp/gpytorch#1630] [Q&A]
I am trying to apply multi-objective Bayesian optimization, I have 2 objective functions that I am trying to approximate using GP. Each objective function can be approximated using different but somehow related data. For instance, if the first objective function is fitted using data x, then the second objective function is fitted using data x multiplied by constant array (c), such as x*c. I am wondering if there is a way to fit both objectives using multi-tasks multivariate GP, or its better to fit each objective separately using multi-dimensional GP? please note that I have 200 observations and each observation consists of 14 variables . Thanks

**Comment by Balandat** [comment]
Is the relationship indeed deterministic as in your example? If so there shouldn't be any reason to have two models, just apply the proper feature transformation and stick it into a single one. 

Are the features living in the same domain? If they do a multi-task model would make sense. If not I would start with independent models. 

If you wanted to get fancy you could also try to a couple of feature extractors on each input that map to some shared latent space on which you define a MTGP model. But I'd start simple.

**Reply by lovecodng** [reply]
Thank you for your reply, I will try with transformation. However, there is a library called botorch I think its based on gpytorch and offer MOBO, which could save me time. Do you recommend it?

**Reply by eytan** [reply]
@lovecodng, [botorch](https://arxiv.org/pdf/1910.06403.pdf) is @Balandat's baby :)  If you aren't developing any new acquisition functions, you might want t start with Ax, which is an end-user library for Bayesian optimization built on top of BoTorch.  BoTorch provides a lower level framework for BayesOpt research which can be helpful if you are developing new acquisition functions or exotic models.

The default MOBO optimizers + models should work quite well out of the box.

There is a tutorial on how to do MOBO w/ Ax here:
  https://ax.dev/tutorials/multiobjective_optimization.html

and it should be possible to plug in a custom GPyTorch model by following this example:
  https://botorch.org/tutorials/custom_botorch_model_in_ax

One risk of using a MTGP is that if the function behavior is quite different (e.g., the lengthscales of each task are different are you are using an ICM kernel), this can result in worse performance than using independent GPs.... It sounds like in your case though, one variable is mechanically some linear combination of the other—in which case why not just use single objective BayesOpt, since there would be no tradeoff?

**Reply by lovecodng** [reply]
Thank you for your reply, The process is a little complicated where I need to modify the BO so I think BoTorch is the most suitable for me. However, do you recommend it to optimize discrete functions?

**Reply by Balandat** [reply]
> However, do you recommend it to optimize discrete functions?

Are all the inputs discrete? Are they ordinal or categorical parameters? How many values can they take on? 

In general there isn't a fundamental problem with optimizing discrete functions, but depending on what the search space looks like the standard models may not make a lot of sense and you'll want to use sth like a `MixedSingleTaskGP` if you have categorical parameters. You will also need to do the acquisition function optimization somewhat differently, e.g. via `optimize_acqf_mixed` or `optimize_acqf_discrete` (which in its currently implementation, depending on the cardinality of the discrete part of the search space, may not scale).
